newIDE/app/src/KeyboardShortcuts/ShortcutsList.js

```diff
@@ -0,0 +1,135 @@
+// @flow
+import * as React from 'react';
+import { t, Trans } from '@lingui/macro';
+import { type I18n } from '@lingui/core';
+import List from '@material-ui/core/List';
+import Text from '../UI/Text';
+import DetectShortcutDialog from './DetectShortcutDialog';
+import { Line } from '../UI/Grid';
+import RaisedButton from '../UI/RaisedButton';
+import { type ShortcutMap } from './DefaultShortcuts';
+import { getShortcutDisplayName } from './index';
+import Window from '../Utils/Window';
+import defaultShortcuts from '../KeyboardShortcuts/DefaultShortcuts';
+import ShortcutsListRow from './ShortcutsListRow';
+import commandsList, {
+  type CommandName,
+  commandAreas,
+} from '../CommandPalette/CommandsList';
+
+/**
+ * Sorts all commands into an object keyed by area name, and also creates a
+ * reverse mapping from shortcut string to list of commands with that shortcut.
+ */
+const sortCommandsIntoAreasAndGetReverseMap = (
+  userShortcutMap: ShortcutMap
+) => {
+  const areaWiseCommands = {};
+  const shortcutStringToCommands: { [string]: Array<CommandName> } = {};
+  Object.keys(commandsList)
+    .filter(name => !commandsList[name].noShortcut)
+    .forEach(name => {
+      // Sort commands by area
+      const areaName = commandsList[name].area;
+      if (!areaWiseCommands[areaName]) areaWiseCommands[areaName] = [];
+      areaWiseCommands[areaName].push(name);
+
+      // Add to shortcut-command mapping
+      const userShortcut = userShortcutMap[name];
+      const defaultShortcut = defaultShortcuts[name] || '';
+      const shortcutString = userShortcut || defaultShortcut;
+      if (shortcutString === '') return;
+      shortcutStringToCommands[shortcutString] = (
+        shortcutStringToCommands[shortcutString] || []
+      ).concat(name);
+    });
+
+  return [areaWiseCommands, shortcutStringToCommands];
+};
```

Comment:
I've extracted the area-wise sorting and shortcut->commands mapping generation into this function. It creates two objects `areaWiseCommands` and `shortcutStringToCommands`, and then proceeds to fill them with a single loop over the commands list, and works properly.

The only thing I'm thinking about is that this function seems to do two quite unrelated things at once. Separating into two functions is cleaner but then we'll have to loop over the commands list twice - complexity is still linear, so should I split the function?